doc/man1/roardtmf.1
.\" roardtmf.1:

.TH "roardtmf" "1" "February 2010" "RoarAudio" "System User's Manual: roardtmf"

.SH NAME

roardtmf \- Generate DTMF (telephone touch tones) for RoarAudio

.SH SYNOPSIS

roardtmf [OPTIONS]... PHONE NUMBER

.SH DESCRIPTION

This tool generates DTMF (telephone touch tones) signals and send them to
the RoarAudio sound server.

In addition to the DTMF digits (0-9, A-D, * and #) space, period (.) and plus (+)
are accepted and ignored so you can copy and past the phone number from any application
easily.

.SH "OPTIONS"

.TP
\fB--server SERVER\fR
Set server hostname

.TP
\fB--rate   RATE\fR
Set sample rate

.TP
\fB--help\fR
Show this help

.SH "SEE ALSO"
\fBroartips\fR(7),
\fBlibroar\fR(7),
\fBRoarAudio\fR(7).

.SH "HISTORY"

For history information see \fBRoarAudio\fR(7).

.\" ll


ChangeLog
v. 0.3beta5 - ?
	* added tool roardtmf
	* Support to store server address in X11's root window propertys
	* some small fixes
	* wrote a lot more libroarpulse code, some async API clients work now
	* added compbins for esddsp, artsdsp, audiooss, padsp
	* splited libroarpulse and libroarpulse-simple
	* upgraded some code to use VIO
	* more updates for Debian
	* fixed init script for Debian and Debian like systems

v. 0.3beta4 - Sat Feb 06 2010 26:02 CET
	* Applyed patches for Debian
	* updated roarify to load libs of correct version
	* updated list of supported versions (complibs)
	* added header with common units
	* added support for PulseAudio Simple protocol
	* added support to libroaross to emulate DMX4Linux (write only)
	* added DSTR driver to roard
	* updated librarys' soname
	* do not link OpenSSL by default (licence incompatible?)
	* support for dynamic lib loading, plugin interface
	* support to for registering DSTR types
	* moved WinSock things out of main vio.[ch]
	* support auto-tool --prefix*=*

v. 0.3beta3 - Fri Jan 22 2010 22:17 CET
	* added some aRts/NAS/YIFF/PulseAudio tools to compbins
	* added support for comp libs to roar-config
	* added support for uni-directional streams in roarinterconnect
	* write much better support for config, including codec config
	* reduced the TCP overhead produced by speex
	* added support for speex preprocessor
	* get names of complibs from table, do not guess them
	* a lot small fixes
	* moved most of the mixer stuff out of roard into libroardsp
	* added flags immutable and enhance
	* test for signal.h
	* Corrected (DSTR) HTTP Protocol support so servers without headerflush
	  should now work, too.
	* added protocol support for icy
	* Added version numbers to lib names
	* updated manpages
	* wrote OSS emulation
	* fixed test scripts

v. 0.3beta2 - Mon Oct 05 2009 19:48 CEST
	* Added prethru flag
	* Added support to read info about client's network node
	* Ported libroaresd, libroarartsc
	* Fixed some problems with EsounD emulation
	* added support for compatibility binarys
	* added poll() support to libroarsndio
	* reject streams of unsupported codec

v. 0.3beta1 - Sat Sep 26 2009 25:50 CEST
	* incremented nummber of streams per client fron 4 to 8
	* RIFF WAVE codecfilter uses virtual streams
	* another set of changed names of parameters in roarctl
	* Interface to read byteorder of a client from roard
	* implemented a 'state of stream' var
	* cleanup, small fixes
	* added VIO interface to transcoder interface
	* ask codecfilter of parent in case of new child stream
	* added 'subsystem' complex: multi subsystem data on single streams
	* added basic /Radio Data and Trasmitter Control System/.
	* better support for stereo upmixing
	* make audio group a optional thing
	* better support for win32
	* wrote WMM driver
	* better multi protocol support
	* Support for server side EsounD emulation on win32

v. 0.3beta0 - Sun Sep 06 2009 22:42 CEST
	* added new stream directions for MIDI and light control
	* cleanup of client stream struct
	* added basic light control subsystem
	* added DMX512 driver
	* wrote basic MIDI subsystem
	* added flag 'mute'
	* added flag 'mmap'
	* added lib libroarmidi and libroarlight
	* added OpenSLP support
	* rewote fillmixbuer2 code
	* added a roar_conv2() with a more clear API
	* added support to write pidfile
	* added support for encoding 32 bit Vorbis streams
	* added transcoding features in libroardsp
	* added tool 'roarphone'
	* added tool 'roarinterconnect'
	* added simple raw data subsystem
	* added working support for THRU streams
	* changed names of parameters in roarctl
	* added lib libroareio - Extended IO for RoarAudio
	* Update order of default drivers, now it is: oss, ao, esd, null
	* cleanup of old sysio driver interface
	* added server side EsounD emulation
	* added multi homed support to server
	* fixed a long outstanding bug with pa_path_get_filename()

v. 0.2 - Thu May 21 2009 21:00
	* do a some more testings for functions not existing on win32
	* init winsock
	* done a lot fixes in order to get libroar and some clients work on win32
	* done a lot cleanup on configure script

v. 0.2rc2 - Sun May 17 2009 25:44
	* corrected check for GNU peercred
	* added basic logging support to configure
	* fixed plugin configures
	* added and fixed semi-install target on plugins
	* call plugin configures from main configure
	* include plugins in Makefile known to work from configure
	* fixed Makefiles of plugins: added -I../../include, -L../../lib
	* Make use of pkg-config in plugin configures
	* corrected possition of -ls on linker commands in most Makefiles
	* added more explicide linker commands so most things now build
	  with classic BSD make.
	* devided into components: libs, roard, clients, doc
	* add toption to set roard's cycle freq
	* moved a lot things in plugin's configure scripts to a commone one

v. 0.2rc1 - Wed May 13 2009 12:05
	* Started to port to OpenSolaris
	* fixed bug in default based sockets, make HTTP VIO from non default port work
	* got libroarsndio build under OpenBSD 4.5
	* do not depend tests in configure on roaraudio.h: fixes some checks on *BSD
	* renamed member 'fh' of connection struct to '__fh', this breaks
	  all code depending on it so we see how it behaves in case
	  we replace this with a real VIO object in near future
	* started VIO 2popen interface
	* fixed exec() on systems with sizeof(int) != sizeof(char*)
	* do not do unnessesery calls to fdatasync() on sockets
	* link roarvorbis against libogg and libvorbis
	* remove need of GNU/Tar like -z on remote system from send-roaraudio
	* added support for hwmixer flag on OSS and sndio driver
	* fiexed some typos

v. 0.2beta3 - Fri May 01 2009 15:15
	* added esdfilt options to roarfilt, esdplay options to roarcatplay
	* added short options to roarcat/mon/filt as stated in the manpage
	* updated manpages
	* also build libroar{,dsp}.a
	* Changed VIO CTLs in a binary inkompatible way
	* added support for autoconf flag to OSS driver

roarclients/roardtmf.c
//roardtmf.c:

/*
 *      Copyright (C) Philipp 'ph3-der-loewe' Schafft - 2008-2010
 *
 *  This file is part of roarclients a part of RoarAudio,
 *  a cross-platform sound system for both, home and professional use.
 *  See README for details.
 *
 *  This file is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 3
 *  as published by the Free Software Foundation.
 *
 *  RoarAudio is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this software; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 51 Franklin Street, Fifth Floor,
 *  Boston, MA 02110-1301, USA.
 *
 */

#include <roaraudio.h>

#ifdef ROAR_HAVE_LIBM
#include <math.h>

// in ms:
#define ON_TIME  (180)
#define OFF_TIME ( 80)
#define SUM_TIME (ON_TIME+OFF_TIME)

struct tone {
 char c;
 float f0;
 float f1;
} g_tones[] = {
 {'1', 697, 1209},
 {'2', 697, 1336},
 {'3', 697, 1477},
 {'A', 697, 1633},

 {'4', 770, 1209},
 {'5', 770, 1336},
 {'6', 770, 1477},
 {'B', 770, 1633},

 {'7', 852, 1209},
 {'8', 852, 1336},
 {'9', 852, 1477},
 {'C', 852, 1633},

 {'*', 941, 1209},
 {'0', 941, 1336},
 {'#', 941, 1477},
 {'D', 941, 1633},

 {0, -1, -1}
};

void usage (void) {
 printf("roardtmf [OPTIONS]... PHONE NUMBER\n");

 printf("\nOptions:\n\n");

 printf("  --server SERVER    - Set server hostname\n"
        "  --rate   RATE      - Set sample rate\n"
//        "  --bits   BITS      - Set bits per sample\n"
//        "  --codec  CODEC     - Set the codec\n"
        "  --help             - Show this help\n"
       );

}

int calc (int16_t * samples, size_t len, int rate, char c) {
 struct tone * ct = NULL;
 int i;
 float t;
 float t_inc = 1./rate;
 float t_max = ON_TIME / 1000.;
 float fc0, fc1;

// printf("calc(*): t_inc=%f, t_max=%f\n", t_inc, t_max);

 if ( c >= 'a' )
  c -= 'a' - 'A';

 for (i = 0; g_tones[i].c != 0; i++) {
  if ( g_tones[i].c == c ) {
   ct = &(g_tones[i]);
   break;
  }
 }

// printf("calc(*): ct=%p\n", ct);

 if ( ct == NULL )
  return -1;

 fc0 = 2 * M_PI * ct->f0;
 fc1 = 2 * M_PI * ct->f1;

// printf("fc0=%f, fc1=%f\n", fc0, fc1);

 memset(samples, 0, len);

 for (i = 0, t = 0; t < t_max; t += t_inc, i++) {
//  printf("i=%i, t=%f\n", i, t);
  samples[i] = (sinf(fc0*t) + sinf(fc1*t))*8192.0;
 }

 return 0;
}

int main (int argc, char * argv[]) {
 int    rate     = ROAR_RATE_DEFAULT;
 int    bits     = ROAR_BITS_DEFAULT;
 int    codec    = ROAR_CODEC_DEFAULT;
 char * server   = NULL;
 char * k;
 int    i;
 char * name = "roardtmf";
 struct roar_vio_calls stream;
 char * tones = NULL;
 void * buf;
 size_t samples;
 size_t len;
 char   c;

 for (i = 1; i < argc; i++) {
  k = argv[i];

  if ( !strcmp(k, "--server") || !strcmp(k, "-s") ) {
   server = argv[++i];
  } else if ( !strcmp(k, "--rate") || !strcmp(k, "-r") ) {
   rate = atoi(argv[++i]);
  } else if ( !strcmp(k, "--bits") ) {
   bits = atoi(argv[++i]);
  } else if ( !strcmp(k, "--codec") ) {
   codec = roar_str2codec(argv[++i]);
  } else if ( !strcmp(k, "--help") ) {
   usage();
   return 0;
  } else if ( tones == NULL ) {
   tones = argv[i++];
  } else {
   fprintf(stderr, "Error: unknown argument: %s\n", k);
   usage();
   return 1;
  }
 }

 if ( tones == NULL ) {
  return 1;
 }

 if ( codec != ROAR_CODEC_DEFAULT )
  return 2;
 if ( bits  != 16 )
  return 3;

 samples = SUM_TIME * rate / 1000;

/*
 printf("samples=%llu\n", (long long unsigned int)samples);
 return 0;
*/

 if ( (buf = malloc((len = bits*samples/8))) == NULL )
  return 4;

 if ( roar_vio_simple_stream(&stream, rate, 1, bits, codec, server, ROAR_DIR_PLAY, name) == -1 ) {
  fprintf(stderr, "Error: can not start playback\n");
  return 1;
 }

 while ( (c = *tones) != 0 ) {
  tones++;

  if ( c == ' ' || c == '+' || c == '.' || c == '_' )
   continue;

  if ( calc(buf, len, rate, c) == -1 )
   return 5;
  roar_vio_write(&stream, buf, len);
 }

 roar_vio_close(&stream);

 return 0;
}

#else
int main (void) {
 fprintf(stderr, "Error: No Math library support compiled in.\n");
 return 1;
}
#endif

//ll
